command/agent/exec/exec.go

```diff
@@ -0,0 +1,288 @@
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
+				s.logger.Info("exec server received new token")
+
+				s.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token:           latestToken,
+						ClientUserAgent: pointerutil.StringPtr(useragent.AgentTemplatingString()),
+					},
+				}
+
+				runnerConfig = runnerConfig.Merge(&ctv)
+				var runnerErr error
+				s.runner, runnerErr = manager.NewRunner(runnerConfig, true)
+				if runnerErr != nil {
+					s.logger.Error("template server failed with new Vault token", "error", runnerErr)
+					continue
+				}
+				go s.runner.Start()
+			}
+
+		case err := <-s.runner.ErrCh:
+			s.logger.Error("template server error", "error", err.Error())
+			s.runner.StopImmediately()
+
+			// Return after stopping the runner if exit on retry failure was specified
+			if s.config.AgentConfig.TemplateConfig != nil && s.config.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+				return fmt.Errorf("template server: %w", err)
+			}
+
+			s.runner, err = manager.NewRunner(runnerConfig, true)
+			if err != nil {
+				return fmt.Errorf("template server failed to create: %w", err)
+			}
+			go s.runner.Start()
+		case <-s.runner.TemplateRenderedCh():
+			// A template has been rendered, figure out what to do
+			s.logger.Debug("template rendered")
+			events := s.runner.RenderEvents()
+
+			// events are keyed by template ID, and can be matched up to the id's from
+			// the lookupMap
+			if len(events) < len(s.lookupMap) {
+				// Not all templates have been rendered yet
+				continue
+			}
+
+			// assume the renders are finished, until we find otherwise
+			doneRendering := true
+			envVarToContents := map[string]string{}
+			for _, event := range events {
+				// This template hasn't been rendered
+				if event.LastWouldRender.IsZero() {
+					doneRendering = false
+				} else {
+					// TODO: check for duplicates?
+					for _, tcfg := range event.TemplateConfigs {
+						envVarToContents[*tcfg.MapToEnvironmentVariable] = string(event.Contents)
+					}
+				}
+			}
+
+			if doneRendering {
+				s.logger.Debug("done rendering templates/detected change, bouncing process")
+				if err := s.bounceCmd(envVarToContents); err != nil {
+					return fmt.Errorf("unable to bounce command: %w", err)
+				}
+			}
+		case exitCode := <-s.exitCh:
+			// process exited on its own
+			return &ProcessExitError{ExitCode: exitCode}
+		}
+	}
+}
+
+func (s *Server) bounceCmd(newEnvVars map[string]string) error {
+	s.procLock.Lock()
+	defer s.procLock.Unlock()
+
+	switch s.config.AgentConfig.Exec.RestartOnSecretChanges {
+	case "always":
+		if s.procStarted {
+			// process is running, need to kill it first
+			s.logger.Info("stopping process", "process_id", s.proc.Pid())
+			s.proc.Stop()
+		}
+	case "never":
+		if s.procStarted {
+			s.logger.Info("detected update, but not restarting process", "process_id", s.proc.Pid())
+			return nil
+		}
+	}
```

Comment:
don't forget a default case

Reply:
TODO: add error in default case (it should never happen anyway)